File: sdk/typescript/src/extract.ts
/**
 * Extraction pipeline: extract structured facts from conversation messages.
 * Matches the Go proxy/extract.go logic exactly.
 */

import { createHash, randomUUID } from 'crypto';
import type { Embedder } from './embedder.js';
import { cosineSimilarity, dimensionMatch } from './search.js';
import { defaultContentKey, type Store } from './store.js';
import type { Fact, FactFilter } from './types.js';

/** Trivial conversation patterns that should not trigger extraction. */
const TRIVIAL_PATTERNS = new Set([
  'thanks', 'thank you', 'ok', 'okay', 'got it', 'sure',
  'hello', 'hi', 'hey', 'bye', 'goodbye', 'good morning',
  'good night', 'good evening', 'good afternoon',
  'cool', 'great', 'awesome', 'nice',
  'understood', 'roger', 'ack', 'k', 'kk', 'lol', 'haha',
]);

/** Valid tags for facts. */
const VALID_TAGS = new Set([
  'skill', 'preference', 'relationship', 'medical',
  'location', 'work', 'hobby', 'personal', 'financial', 'other',
]);

/** Semantic dedup threshold — 0.92 cosine similarity. */
const SEMANTIC_DEDUP_THRESHOLD = 0.92;

/** Reinforcement count threshold for promotion to high significance. */
const PROMOTION_THRESHOLD = 5;

/** JSON shape returned by the extraction LLM. */
interface ExtractedFact {
  content: string;
  type: string;
  significance: number;
  tag: string;
  slot: string;
  reference_time: string;
  confidence: number | null;
}

/** Message shape for extraction. */
export interface ExtractionMessage {
  role: string;
  content: string;
}

/**
 * Check if a set of messages is trivial (greetings, acknowledgments).
 */
export function isTrivialTurn(messages: ExtractionMessage[]): boolean {
  if (messages.length === 0) return true;

  for (const msg of messages) {
    if (msg.role !== 'user') continue;

    let content = msg.content.toLowerCase().trim();
    content = content.replace(/[^a-z0-9 ]/g, '').trim();

    if (content === '') continue;

    if (!TRIVIAL_PATTERNS.has(content)) {
      return false;
    }
  }

  return true;
}

/**
 * Build the extraction prompt. Matches Go exactly.
 */
function buildExtractionPrompt(): string {
  const now = new Date();
  const today = now.toISOString().slice(0, 10);
  const yesterday = new Date(now.getTime() - 86400000).toISOString().slice(0, 10);

  return `You are a knowledge extraction engine. Today's date is ${today}.

Extract facts from this conversation. Return a JSON array. Each fact:
{
  "content": "the fact as a clear statement about the user",
  "type": "identity|event|pattern",
  "significance": 1-10,
  "tag": "category label",
  "slot": "semantic slot name (e.g. location, job, diet, name, email, relationship, preference)",
  "reference_time": "ISO date if time-bound, empty string if not",
  "confidence": 0.0-1.0
}

Rules:
- "identity" = enduring truths (preferences, attributes, relationships)
- "event" = things that happened at a specific time
- "pattern" = behavioral tendencies observed across the conversation
- "tag" = a category label. Use one of: skill, preference, relationship, medical, location, work, hobby, personal, financial, or other
- "slot" = the semantic slot this fact fills. Use: location, job, diet, name, email, relationship, preference, medical, hobby, skill, or other
- "reference_time" = ISO 8601 date (YYYY-MM-DD) for time-bound facts, empty string otherwise
- "confidence" = how confident you are (1.0 = explicitly stated by user, 0.5 = inferred, 0.0 = guessing)
- Resolve relative dates: "today" → "${today}", "yesterday" → "${yesterday}"
- Significance: 10 = life-critical (allergies, medical), 7-9 = important (job, location), 4-6 = moderate, 1-3 = trivial (lunch, weather)
- Skip greetings, filler, "thank you", and trivial exchanges
- If nothing is worth extracting, return []

Return ONLY the JSON array, no other text.`;
}

/**
 * Parse LLM response into extracted facts.
 * Resilient to markdown code blocks and surrounding text.
 */
function parseExtractionResponse(content: string): ExtractedFact[] {
  content = content.trim();

  // Try direct parse.
  try {
    return JSON.parse(content) as ExtractedFact[];
  } catch { /* continue */ }

  // Try stripping code fences.
  const stripped = stripCodeFences(content);
  if (stripped !== content) {
    try {
      return JSON.parse(stripped) as ExtractedFact[];
    } catch { /* continue */ }
  }

  // Try finding JSON array within the text.
  const start = content.indexOf('[');
  const end = content.lastIndexOf(']');
  if (start >= 0 && end > start) {
    const candidate = content.slice(start, end + 1);
    try {
      return JSON.parse(candidate) as ExtractedFact[];
    } catch { /* continue */ }
  }

  throw new Error(`Could not parse JSON array from response: ${content.slice(0, 100)}`);
}

/**
 * Strip markdown code block wrappers.
 */
function stripCodeFences(s: string): string {
  s = s.trim();

  if (s.startsWith('```')) {
    const idx = s.indexOf('\n');
    if (idx >= 0) {
      s = s.slice(idx + 1);
    } else {
      s = s.replace(/^```json/, '').replace(/^```/, '');
    }
  }

  if (s.endsWith('```')) {
    s = s.slice(0, -3);
  }

  return s.trim();
}

/**
 * Validate and filter extracted facts.
 */
function validateExtraction(facts: ExtractedFact[]): ExtractedFact[] {
  const valid: ExtractedFact[] = [];

  for (const f of facts) {
    const content = (f.content ?? '').trim();
    if (content === '') continue;
    if (content.length > 500) continue;

    if (f.reference_time) {
      const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
      if (!dateRegex.test(f.reference_time)) {
        f.reference_time = '';
      } else {
        const parsed = new Date(f.reference_time);
        if (isNaN(parsed.getTime())) {
          f.reference_time = '';
        }
      }
    }

    if (f.confidence !== null && f.confidence !== undefined) {
      if (f.confidence < 0) f.confidence = 0;
      if (f.confidence > 1) f.confidence = 1;
    }

    let tag = (f.tag ?? '').toLowerCase().trim();
    if (tag && !VALID_TAGS.has(tag)) {
      tag = 'other';
    }

    valid.push({
      ...f,
      content,
      tag,
    });
  }

  return valid;
}

/**
 * Resolve fact type string to a valid type.
 */
function resolveFactType(t: string): 'identity' | 'event' | 'pattern' {
  switch ((t ?? '').toLowerCase().trim()) {
    case 'event':
      return 'event';
    case 'pattern':
      return 'pattern';
    case 'identity':
    default:
      return 'identity';
  }
}

/**
 * Clamp significance to [1, 10], defaulting to 5.
 */
function clampSignificance(v: number): number {
  if (!v || v < 1 || v > 10) return 5;
  return v;
}

/**
 * Get confidence value, defaulting to 0.8 if null.
 */
function confidenceValue(v: number | null | undefined): number {
  if (v === null || v === undefined) return 0.8;
  return v;
}

/**
 * Compute TTL expiry based on significance. Matches Go TTLForSignificance.
 */
function ttlForSignificance(sig: number): string | null {
  if (sig >= 10) return null; // never expires
  const now = Date.now();
  if (sig >= 5) {
    return new Date(now + 30 * 24 * 60 * 60 * 1000).toISOString(); // ~1 month
  }
  return new Date(now + 7 * 24 * 60 * 60 * 1000).toISOString(); // ~1 week
}

/**
 * Call an LLM via fetch for extraction. Supports OpenAI-compatible APIs.
 */
export async function callLLM(
  apiKey: string,
  model: string,
  systemPrompt: string,
  userContent: string,
  provider: string
): Promise<string> {
  let url: string;
  let headers: Record<string, string>;
  let body: string;

  if (provider === 'gemini') {
    url = `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent?key=${apiKey}`;
    headers = { 'Content-Type': 'application/json' };
    body = JSON.stringify({
      systemInstruction: { parts: [{ text: systemPrompt }] },
      contents: [{ role: 'user', parts: [{ text: userContent }] }],
      generationConfig: { temperature: 0.1 },
    });
  } else if (provider === 'anthropic') {
    url = 'https://api.anthropic.com/v1/messages';
    headers = {
      'Content-Type': 'application/json',
      'x-api-key': apiKey,
      'anthropic-version': '2023-06-01',
    };
    body = JSON.stringify({
      model,
      max_tokens: 2048,
      system: systemPrompt,
      messages: [{ role: 'user', content: userContent }],
    });
  } else {
    url = 'https://api.openai.com/v1/chat/completions';
    headers = {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${apiKey}`,
    };
    body = JSON.stringify({
      model,
      max_tokens: 2048,
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: userContent },
      ],
    });
  }

  let lastErr: Error | null = null;
  const maxRetries = 3;
  let backoff = 1000;

  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const resp = await fetch(url, { method: 'POST', headers, body });

      if (!resp.ok) {
        const respBody = await resp.text();
        const errMsg = `LLM API error: HTTP ${resp.status} ${respBody}`;
        lastErr = new Error(errMsg);

        const retryable =
          resp.status === 429 ||
          resp.status >= 500;

        if (!retryable || attempt === maxRetries) throw lastErr;

        await new Promise((r) => setTimeout(r, backoff));
        backoff *= 2;
        continue;
      }

      const data = await resp.json() as any;

      if (provider === 'gemini') {
        return data.candidates?.[0]?.content?.parts?.[0]?.text ?? '';
      }

      if (provider === 'anthropic') {
        const blocks = data.content as any[];
        const textBlock = blocks?.find((b: any) => b.type === 'text');
        return textBlock?.text ?? '';
      }

      return data.choices?.[0]?.message?.content ?? '';
    } catch (err: any) {
      lastErr = err;
      if (attempt === maxRetries) throw lastErr;

      await new Promise((r) => setTimeout(r, backoff));
      backoff *= 2;
    }
  }

  throw lastErr ?? new Error('LLM call failed');
}

/**
 * Find the most semantically similar existing fact above the dedup threshold.
 */
function findSemanticMatch(
  embedding: number[],
  existingFacts: Fact[]
): Fact | null {
  let best: Fact | null = null;
  let bestScore = 0;

  for (const f of existingFacts) {
    if (!f.embedding || f.embedding.length === 0) continue;
    const score = cosineSimilarity(embedding, f.embedding);
    if (score > bestScore) {
      bestScore = score;
      best = f;
    }
  }

  if (bestScore >= SEMANTIC_DEDUP_THRESHOLD) return best;
  return null;
}

/**
 * Find existing current identity facts that occupy the same slot.
 */
function findSlotConflicts(slot: string, existing: Fact[]): Fact[] {
  if (!slot) return [];
  return existing.filter(
    (f) =>
      f.slot === slot &&
      f.temporalStatus === 'current' &&
      f.factType === 'identity'
  );
}

/**
 * Run the full extraction pipeline:
 * 1. Trivial turn detection
 * 2. LLM extraction call
 * 3. Parse and validate
 * 4. Embed extracted facts
 * 5. Dedup (content key + semantic)
 * 6. Slot conflict resolution
 * 7. TTL assignment
 * 8. Persist
 */
export async function runExtraction(
  store: Store,
  embedder: Embedder | null,
  messages: ExtractionMessage[],
  entityUuid: string,
  apiKey: string,
  llmModel: string,
  llmProvider: string
): Promise<{ inserted: number; reinforced: number }> {
  if (messages.length === 0) return { inserted: 0, reinforced: 0 };
  if (isTrivialTurn(messages)) return { inserted: 0, reinforced: 0 };

  const userMessages = messages.filter((m) => m.role === 'user');
  if (userMessages.length === 0) return { inserted: 0, reinforced: 0 };

  // Build transcript from user messages only.
  const transcript = userMessages.map((m) => `${m.role}: ${m.content}`).join('\n');

  const systemPrompt = buildExtractionPrompt();

  let responseContent: string;
  try {
    responseContent = await callLLM(apiKey, llmModel, systemPrompt, transcript, llmProvider);
  } catch (err) {
    console.warn('[memg] extraction LLM call failed:', err);
    return { inserted: 0, reinforced: 0 };
  }

  let extracted: ExtractedFact[];
  try {
    extracted = parseExtractionResponse(responseContent);
  } catch (err) {
    console.warn('[memg] extraction parse failed:', err);
    return { inserted: 0, reinforced: 0 };
  }

  extracted = validateExtraction(extracted);
  if (extracted.length === 0) return { inserted: 0, reinforced: 0 };

  // Build facts.
  const embeddingModelName = embedder?.modelName() ?? '';
  const facts: Fact[] = [];
  const contents: string[] = [];

  for (const ef of extracted) {
    const factType = resolveFactType(ef.type);
    const significance = clampSignificance(ef.significance);

    const fact: Fact = {
      uuid: randomUUID(),
      content: ef.content,
      factType,
      temporalStatus: 'current',
      significance,
      contentKey: defaultContentKey(ef.content),
      tag: (ef.tag ?? '').toLowerCase().trim(),
      slot: (ef.slot ?? '').toLowerCase().trim(),
      confidence: confidenceValue(ef.confidence),
      embeddingModel: embeddingModelName,
      sourceRole: 'user',
      reinforcedCount: 0,
      recallCount: 0,
      expiresAt: ttlForSignificance(significance) ?? undefined,
      referenceTime: undefined,
    };

    if (ef.reference_time) {
      const parsed = new Date(ef.reference_time);
      if (!isNaN(parsed.getTime())) {
        fact.referenceTime = ef.reference_time;
      }
    }

    facts.push(fact);
    contents.push(ef.content);
  }

  // Embed all fact contents in a single batch.
  if (embedder) {
    try {
      const embeddings = await embedder.embed(contents);
      for (let i = 0; i < embeddings.length && i < facts.length; i++) {
        facts[i].embedding = embeddings[i];
      }
    } catch (err) {
      console.warn('[memg] extraction embed failed, continuing without embeddings:', err);
    }
  }

  // Load existing facts for dedup and conflict detection.
  const existingFacts = await store.listFactsForRecall(
    entityUuid,
    { statuses: ['current'], excludeExpired: true },
    0
  );

  let inserted = 0;
  let reinforced = 0;

  for (const fact of facts) {
    // Step 1: Exact dedup by content key.
    const existing = await store.findFactByKey(entityUuid, fact.contentKey);
    if (existing) {
      await reinforceAndPromote(store, existing, fact.expiresAt ?? null);
      reinforced++;
      continue;
    }

    // Step 2: Slot conflict resolution.
    if (
      fact.slot &&
      fact.factType === 'identity' &&
      fact.temporalStatus === 'current'
    ) {
      const conflicts = findSlotConflicts(fact.slot, existingFacts);
      for (const conflict of conflicts) {
        await store.updateTemporalStatus(conflict.uuid, 'historical');
        conflict.temporalStatus = 'historical';
      }
    }

    // Step 3: Semantic dedup.
    if (fact.embedding && fact.embedding.length > 0 && existingFacts.length > 0) {
      const match = findSemanticMatch(fact.embedding, existingFacts);
      if (match) {
        await reinforceAndPromote(store, match, fact.expiresAt ?? null);
        reinforced++;
        continue;
      }
    }

    // Step 4: Insert the new fact.
    await store.insertFact(entityUuid, fact);
    inserted++;
  }

  return { inserted, reinforced };
}

/**
 * Reinforce an existing fact and promote to high significance if warranted.
 */
async function reinforceAndPromote(
  store: Store,
  existing: Fact,
  newExpiresAt: string | null
): Promise<void> {
  const newCount = existing.reinforcedCount + 1;
  const shouldPromote =
    newCount >= PROMOTION_THRESHOLD && existing.significance < 10;

  let expiresAt = newExpiresAt;
  if (shouldPromote) {
    expiresAt = null; // high-significance facts never expire
  }

  await store.reinforceFact(existing.uuid, expiresAt);

  if (shouldPromote) {
    await store.updateSignificance(existing.uuid, 10);
  }
}

File: sdk/python/memg/extract.py
"""Extraction pipeline: trivial detection, LLM extraction, validation, dedup."""

from __future__ import annotations

import hashlib
import json
import logging
import re
import time
import unicodedata
from datetime import datetime, timedelta, timezone
from typing import List, Optional

from .search import cosine_similarity
from .store import Store
from .types import Fact, FactFilter

logger = logging.getLogger("memg")

TRIVIAL_PATTERNS = [
    "thanks", "thank you", "ok", "okay", "got it", "sure",
    "hello", "hi", "hey", "bye", "goodbye", "good morning",
    "good night", "good evening", "good afternoon",
    "cool", "great", "awesome", "nice",
    "understood", "roger", "ack", "k", "kk", "lol", "haha",
]

VALID_TAGS = {
    "skill", "preference", "relationship", "medical", "location",
    "work", "hobby", "personal", "financial", "other",
}

SEMANTIC_DEDUP_THRESHOLD = 0.92
PROMOTION_THRESHOLD = 5
SIGNIFICANCE_HIGH = 10
SIGNIFICANCE_MEDIUM = 5


def content_key(content: str) -> str:
    """Compute content key: sha256(lowercase, strip punct, collapse whitespace)[:16].

    First 8 bytes of the hash as hex = 16 character string.
    """
    lower = content.lower()
    cleaned = ""
    for ch in lower:
        if ch.isalpha() or ch.isdigit() or ch.isspace():
            cleaned += ch
    normalized = " ".join(cleaned.split())
    h = hashlib.sha256(normalized.encode("utf-8")).digest()
    return h[:8].hex()


def ttl_for_significance(significance: int) -> Optional[datetime]:
    """Compute expiry time based on significance level."""
    if significance >= SIGNIFICANCE_HIGH:
        return None
    now = datetime.now(timezone.utc)
    if significance >= SIGNIFICANCE_MEDIUM:
        return now + timedelta(days=30)
    return now + timedelta(days=7)


def is_trivial_turn(messages: List[dict]) -> bool:
    """Check if messages contain only trivial content."""
    if not messages:
        return True

    for msg in messages:
        role = msg.get("role", "")
        if role != "user":
            continue
        raw = msg.get("content", "")
        if not raw:
            continue
        cleaned = raw.lower().strip()
        cleaned = "".join(
            ch for ch in cleaned
            if ch.isalpha() or ch.isdigit() or ch == " "
        ).strip()
        if not cleaned:
            continue

        is_triv = any(cleaned == p for p in TRIVIAL_PATTERNS)
        if not is_triv:
            return False

    return True


def build_extraction_prompt() -> str:
    """Build the LLM extraction prompt with today's date."""
    now = datetime.now(timezone.utc)
    today = now.strftime("%Y-%m-%d")
    yesterday = (now - timedelta(days=1)).strftime("%Y-%m-%d")

    return (
        f"You are a knowledge extraction engine. Today's date is {today}.\n\n"
        "Extract facts from this conversation. Return a JSON array. Each fact:\n"
        "{\n"
        '  "content": "the fact as a clear statement about the user",\n'
        '  "type": "identity|event|pattern",\n'
        '  "significance": 1-10,\n'
        '  "tag": "category label",\n'
        '  "slot": "semantic slot name (e.g. location, job, diet, name, email, relationship, preference)",\n'
        '  "reference_time": "ISO date if time-bound, empty string if not",\n'
        '  "confidence": 0.0-1.0\n'
        "}\n\n"
        "Rules:\n"
        '- "identity" = enduring truths (preferences, attributes, relationships)\n'
        '- "event" = things that happened at a specific time\n'
        '- "pattern" = behavioral tendencies observed across the conversation\n'
        '- "tag" = a category label. Use one of: skill, preference, relationship, medical, location, work, hobby, personal, financial, or other\n'
        '- "slot" = the semantic slot this fact fills. Use: location, job, diet, name, email, relationship, preference, medical, hobby, skill, or other\n'
        '- "reference_time" = ISO 8601 date (YYYY-MM-DD) for time-bound facts, empty string otherwise\n'
        '- "confidence" = how confident you are (1.0 = explicitly stated by user, 0.5 = inferred, 0.0 = guessing)\n'
        f'- Resolve relative dates: "today" -> "{today}", "yesterday" -> "{yesterday}"\n'
        "- Significance: 10 = life-critical (allergies, medical), 7-9 = important (job, location), 4-6 = moderate, 1-3 = trivial (lunch, weather)\n"
        "- Skip greetings, filler, \"thank you\", and trivial exchanges\n"
        "- If nothing is worth extracting, return []\n\n"
        "Return ONLY the JSON array, no other text."
    )


def parse_extraction_response(content: str) -> list:
    """Parse JSON array from LLM response, resilient to markdown wrappers."""
    content = content.strip()

    try:
        return json.loads(content)
    except (json.JSONDecodeError, ValueError):
        pass

    stripped = _strip_code_fences(content)
    if stripped != content:
        try:
            return json.loads(stripped)
        except (json.JSONDecodeError, ValueError):
            pass

    start = content.find("[")
    end = content.rfind("]")
    if start >= 0 and end > start:
        candidate = content[start:end + 1]
        try:
            return json.loads(candidate)
        except (json.JSONDecodeError, ValueError):
            pass

    logger.warning("memg extract: could not parse JSON from response: %.100s", content)
    return []


def _strip_code_fences(s: str) -> str:
    s = s.strip()
    if s.startswith("```"):
        idx = s.find("\n")
        if idx >= 0:
            s = s[idx + 1:]
        else:
            s = s.lstrip("`").lstrip("json").lstrip()
    if s.endswith("```"):
        s = s[:-3]
    return s.strip()


def validate_extraction(facts: list) -> list:
    """Filter and validate extracted facts."""
    valid = []
    for f in facts:
        if not isinstance(f, dict):
            continue
        c = (f.get("content") or "").strip()
        if not c:
            continue
        if len(c) > 500:
            continue

        ref_time = f.get("reference_time", "")
        if ref_time:
            try:
                datetime.strptime(ref_time, "%Y-%m-%d")
            except ValueError:
                f["reference_time"] = ""

        confidence = f.get("confidence")
        if confidence is not None:
            try:
                confidence = float(confidence)
                confidence = max(0.0, min(1.0, confidence))
                f["confidence"] = confidence
            except (TypeError, ValueError):
                f["confidence"] = None

        tag = (f.get("tag") or "").lower().strip()
        if tag and tag not in VALID_TAGS:
            tag = "other"
        f["tag"] = tag

        f["content"] = c
        valid.append(f)
    return valid


def _resolve_fact_type(t: str) -> str:
    t = (t or "").lower().strip()
    if t in ("event", "pattern", "identity"):
        return t
    return "identity"


def _clamp_significance(v) -> int:
    try:
        v = int(v)
    except (TypeError, ValueError):
        return SIGNIFICANCE_MEDIUM
    if v < 1 or v > 10:
        return SIGNIFICANCE_MEDIUM
    return v


def _confidence_value(v) -> float:
    if v is None:
        return 0.8
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.8


def extracted_to_facts(
    extracted: list,
    embedding_model: str = "",
    embedder=None,
) -> List[Fact]:
    """Convert validated extraction dicts to Fact objects.

    Optionally embeds fact content using the provided embedder.
    """
    if not extracted:
        return []

    facts = []
    contents = []

    for ef in extracted:
        fact_type = _resolve_fact_type(ef.get("type", ""))
        significance = _clamp_significance(ef.get("significance", 5))

        f = Fact(
            uuid="",
            content=ef["content"],
            fact_type=fact_type,
            temporal_status="current",
            significance=significance,
            tag=(ef.get("tag") or "").lower().strip(),
            slot=(ef.get("slot") or "").lower().strip(),
            confidence=_confidence_value(ef.get("confidence")),
            embedding_model=embedding_model,
            source_role="user",
            content_key=content_key(ef["content"]),
            expires_at=ttl_for_significance(significance),
        )

        ref_time = ef.get("reference_time", "")
        if ref_time:
            try:
                f.reference_time = datetime.strptime(ref_time, "%Y-%m-%d").replace(tzinfo=timezone.utc)
            except ValueError:
                pass

        facts.append(f)
        contents.append(ef["content"])

    if embedder and contents:
        try:
            embeddings = embedder.embed(contents)
            for i, emb in enumerate(embeddings):
                if i < len(facts):
                    facts[i].embedding = emb
        except Exception as e:
            logger.warning("memg extract: embedding failed: %s", e)

    return facts


def persist_facts(
    store: Store,
    entity_uuid: str,
    facts: List[Fact],
) -> tuple:
    """Persist facts with dedup, slot conflict resolution, and promotion.

    Returns (inserted_count, reinforced_count).
    """
    if not facts or not entity_uuid:
        return 0, 0

    existing_facts = store.list_facts_for_recall(
        entity_uuid,
        FactFilter(statuses=["current"], exclude_expired=True),
        limit=500,
    )

    inserted = 0
    reinforced = 0

    for f in facts:
        if not f.content_key:
            f.content_key = content_key(f.content)

        if f.expires_at is None and f.significance < SIGNIFICANCE_HIGH:
            f.expires_at = ttl_for_significance(f.significance)

        # Step 1: Exact dedup by content key
        existing = store.find_fact_by_key(entity_uuid, f.content_key)
        if existing is not None:
            _reinforce_and_promote(store, existing, f.expires_at)
            reinforced += 1
            continue

        # Step 2: Slot conflict resolution for mutable identity facts
        if f.slot and f.fact_type == "identity" and f.temporal_status == "current":
            for ex in existing_facts:
                if (ex.slot == f.slot and
                        ex.temporal_status == "current" and
                        ex.fact_type == "identity"):
                    store.update_temporal_status(ex.uuid, "historical")
                    ex.temporal_status = "historical"

        # Step 3: Semantic dedup by embedding similarity
        if f.embedding and existing_facts:
            similar = _find_semantic_match(f.embedding, existing_facts)
            if similar is not None:
                _reinforce_and_promote(store, similar, f.expires_at)
                reinforced += 1
                continue

        # Step 4: Insert new fact
        store.insert_fact(entity_uuid, f)
        inserted += 1

    return inserted, reinforced


def _reinforce_and_promote(
    store: Store,
    existing: Fact,
    new_expires_at: Optional[datetime],
) -> None:
    """Reinforce an existing fact and promote if threshold is crossed."""
    new_count = existing.reinforced_count + 1
    should_promote = new_count >= PROMOTION_THRESHOLD and existing.significance < SIGNIFICANCE_HIGH

    expires_at = new_expires_at
    if should_promote:
        expires_at = None

    store.reinforce_fact(existing.uuid, expires_at)

    if should_promote:
        store.update_significance(existing.uuid, SIGNIFICANCE_HIGH)
        logger.info(
            "memg extract: promoted fact %s to high significance (reinforced %d times)",
            existing.uuid, new_count,
        )


def _find_semantic_match(
    embedding: List[float],
    candidates: List[Fact],
) -> Optional[Fact]:
    """Find the most similar existing fact by embedding, or None."""
    best = None
    best_score = 0.0

    for f in candidates:
        if not f.embedding:
            continue
        score = cosine_similarity(embedding, f.embedding)
        if score > best_score:
            best_score = score
            best = f

    if best_score >= SEMANTIC_DEDUP_THRESHOLD:
        return best
    return None


def run_extraction(
    store: Store,
    entity_uuid: str,
    messages: List[dict],
    llm_call,
    embedder=None,
) -> tuple:
    """Full extraction pipeline.

    Args:
        store: SQLiteStore instance
        entity_uuid: entity UUID
        messages: list of message dicts with 'role' and 'content'
        llm_call: callable(system_prompt, user_content) -> response_text
        embedder: optional Embedder for fact embedding

    Returns (inserted, reinforced).
    """
    if not messages:
        return 0, 0

    if is_trivial_turn(messages):
        return 0, 0

    user_messages = [m for m in messages if m.get("role") == "user"]
    if not user_messages:
        return 0, 0

    transcript = "\n".join(
        f"{m['role']}: {m['content']}" for m in user_messages
    )

    prompt = build_extraction_prompt()

    try:
        response_text = llm_call(prompt, transcript)
    except Exception as e:
        logger.warning("memg extract: LLM call failed: %s", e)
        return 0, 0

    raw_facts = parse_extraction_response(response_text)
    validated = validate_extraction(raw_facts)
    if not validated:
        return 0, 0

    embedding_model = embedder.model_name() if embedder else ""
    facts = extracted_to_facts(validated, embedding_model, embedder)
    if not facts:
        return 0, 0

    return persist_facts(store, entity_uuid, facts)
